### PR TITLE
use relative path within same context #224

### PR DESCRIPTION
**Pull request by bryanbcook**
References once within the context of the external repo should be relative.

### CODE REVIEW

File: src/Sdk/AzurePipelines/AzureDevops.cs [hunk 1]
```diff
@@ -466,7 +466,7 @@ public class AzureDevops {
         var fileId = templateContext.GetFileId(filename);
         // Read the file
         var finalRepository = afilenameAndRef.Length == 1 ? context.RepositoryAndRef : string.Equals(afilenameAndRef[1], "self", StringComparison.OrdinalIgnoreCase) ? null : context.Repositories.TryGetValue(afilenameAndRef[1], out var ralias) ? ralias : throw new Exception($"Couldn't find repository with alias {afilenameAndRef[1]} in repository resources");
-        var finalFileName = afilenameAndRef.Length == 1 ? RelativeTo(context.CWD ?? ".", filename) : filename;
+        var finalFileName = context.RepositoryAndRef == finalRepository ? RelativeTo(context.CWD ?? ".", filename) : filename;
```
[the remark this answers is not in the document]
Comment (bryanbcook) [reply src/Sdk/AzurePipelines/AzureDevops.cs:469]: Ok, I think the `extends` is a red-herring. I think I have discovered a few additional errors related to `extends` that I'll report separately.

For this issue, my fix states "if we're in a context, resolve paths within that context. If we're in a different context, assume the path is absolute."

What you have written is functionally identical, but with a subtle difference that assumes path usage in the yaml only has a single syntax. If we change the reference to `steps.yml@shared` the error appears.

This probably makes no sense and you would sensibly argue -- why would someone refer to files in their own repository using the alias that was declared in the calling pipeline? 🤪 You're right, it makes no sense, but it doesn't change the fact that it is valid syntax.

(Our business case is we have only one "shared" repo, and teams call into our shared steps using `@shared`. We also have shared templates that make calls back to user-supplied references to `@self`, so to prevent ambiguity the team always scopes their references. As a rule, we generally use absolute paths, but apparently we use relative paths with `@shared` as well)

**Comment by bryanbcook** [comment]
That is an awesome corner case, referring to same repo twice with different aliases! I suppose that could occur if you need to test a template from a specific branch.

I did a quick test in my Azure Pipeline organization, and I don't think the corner case applies.  

```yaml
resources:
  repositories:
  - repository: alias1
    name: 'Pipeline Experiments'
    type: git
    ref: bcook/extends-variables
    
  - repository: alias2
    name: 'Pipeline Experiments'
    type: git
    ref: bcook/extends-variables

jobs:
- template: /devops/pipelines/extends/relative-paths/templates/job.yml@alias1
```
```yaml
# job.yml
jobs:
- job: job1
  steps:
  - template: steps.yml@alias2
```
```yaml
# steps.yml
steps:
- script: exit 0
```

The implementation I provided matches the Azure Pipelines behaviour: if you resolved the template as `@templates` you can use relative paths to other files within the `@templates` context. But if you attempt to use a different context (`@template`) it must be an absolute path.

image
